packages/framework/src/Model/Order/Item/OrderItem.php

```diff
@@ -192,6 +212,42 @@ public function getPriceWithVat(): Money
         return $this->priceWithVat;
     }
 
+    /**
+     * @return \Shopsys\FrameworkBundle\Component\Money\Money|null
+     */
+    public function getTotalPriceWithoutVat(): ?Money
+    {
+        return $this->hasForcedTotalPrice() ? $this->totalPriceWithoutVat : null;
+    }
+
+    /**
+     * @return \Shopsys\FrameworkBundle\Component\Money\Money
+     */
+    public function getTotalPriceWithVat(): Money
+    {
+        return $this->hasForcedTotalPrice() ? $this->totalPriceWithVat : $this->priceWithVat->multiply($this->quantity);
+    }
+
+    /**
+     * The total price property can be used when order item has prices that differ from current price calculation implementation.
+     * Otherwise it should be set to NULL (which means it will be calculated automatically).
+     *
+     * @param \Shopsys\FrameworkBundle\Model\Pricing\Price|null $totalPrice
+     */
+    public function setTotalPrice(?Price $totalPrice): void
+    {
+        $this->totalPriceWithVat = $totalPrice !== null ? $totalPrice->getPriceWithVat() : null;
+        $this->totalPriceWithoutVat = $totalPrice !== null ? $totalPrice->getPriceWithoutVat() : null;
+    }
+
+    /**
+     * @return bool
+     */
+    public function hasForcedTotalPrice(): bool
+    {
+        return $this->totalPriceWithoutVat !== null && $this->totalPriceWithVat !== null;
```

Comment:
there is an unhanded invalid state: One price is not null and one is null. In such case the class will behave as the prices are not forced. But it should throw an exception in my opinion as this case is simply invalid

Reply:
You're right about the invalid state. I let it slide as the state is encapsulated in the entity - the properties can be set only both at once in the `setTotalPrice` method. But I guess as long as the properties are only `protected` and the properties are set using reflection by Doctrine, a bit of defensive programming may be appropriate.

Addressed in 28a65a5a4 using a custom `RuntimeException`. Please note, that we always allow for passing the `$previous` exception as an constructor argument, even if it's unused. I'm keeping this up for consistency.